roster/tests.py

```diff
@@ -1052,3 +1052,76 @@ def test_link_assistant(self) -> None:
         self.assertEqual(alice.assistant.pk, assistant.pk)
         resp = self.assertGetOK("link-assistant")
         self.assertEqual(len(resp.context["form"].fields["student"].queryset), 1)
+
+    def test_cancel_inquiry_sets_status_to_canceled(self):
+        alice = StudentFactory.create()
+        unit = UnitFactory.create()
+        inquiry = UnitInquiry.objects.create(
+            student=alice,
+            unit=unit,
+            action_type="INQ_ACT_UNLOCK",
+            status="INQ_NEW",
+            explanation="Please unlock",
+        )
+        self.login(alice)
+        resp = self.post(
+            "inquiry-cancel",
+            inquiry.pk,
+            follow=True,
+        )
+        inquiry.refresh_from_db()
+        self.assertEqual(inquiry.status, "INQ_CANC")
+        self.assertHas(resp, "Inquiry successfully canceled.")
+
+    def test_only_owner_or_staff_can_cancel(self):
+        alice = StudentFactory.create()
+        bob = StudentFactory.create()
+        staff = UserFactory.create(is_staff=True)
+        unit = UnitFactory.create()
+        inquiry = UnitInquiry.objects.create(
+            student=alice,
+            unit=unit,
+            action_type="INQ_ACT_UNLOCK",
+            status="INQ_NEW",
+            explanation="Please unlock",
+        )
+        # Bob cannot cancel Alice's inquiry
+        self.login(bob)
+        resp = self.post("inquiry-cancel", inquiry.pk, follow=True)
+        self.assertEqual(resp.status_code, 403)
+        # Staff can cancel
+        self.login(staff)
+        resp = self.post("inquiry-cancel", inquiry.pk, follow=True)
+        inquiry.refresh_from_db()
+        self.assertEqual(inquiry.status, "INQ_CANC")
+
+    def test_cancel_button_only_for_pending(self):
+        for status in ["INQ_ACC", "INQ_REJ", "INQ_HOLD", "INQ_CANC"]:
+            alice = StudentFactory.create()
+            unit = UnitFactory.create()
+            UnitInquiry.objects.create(
+                student=alice,
+                unit=unit,
+                action_type="INQ_ACT_UNLOCK",
+                status=status,
+                explanation="Test",
+            )
+            self.login(alice)
+            resp = self.get("inquiry", alice.pk)
+            self.assertNotIn("Cancel", resp.content.decode())
+
+    def test_cannot_cancel_non_pending(self):
+        alice = StudentFactory.create()
+        unit = UnitFactory.create()
+        for status in ["INQ_ACC", "INQ_REJ", "INQ_HOLD", "INQ_CANC"]:
+            inquiry = UnitInquiry.objects.create(
+                student=alice,
+                unit=unit,
+                action_type="INQ_ACT_UNLOCK",
+                status=status,
+                explanation="Test",
+            )
+            self.login(alice)
+            self.post("inquiry-cancel", inquiry.pk, follow=True)
```

Comment:
we should also check the status code is 40x

Reply:
from evan_django_tools there's a shortcut for just `self.assertPost40X` replacing `self.post` for this

althouh really, for a 40X response you probably don't need to inspect the content, the status code is enough in this case imo